compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertInfiniteLoopAroundCoreBlock.cpp

```diff
@@ -0,0 +1,129 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file inserts infinite looping around the `amdaie.core` blocks. This
+// results in the cores running the same program over and over which is useful
+// for measuring performance statistics like latency/throughput, averaged over
+// a certain number of runs, while excluding core reconfiguration overhead.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/AMDAIEOpUtils.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+
+#define DEBUG_TYPE "iree-amdaie-insert-infinite-loop-around-core-block"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+LogicalResult addInfiniteWhileLoopAroundCoreBlock(RewriterBase &rewriter,
+                                                  AMDAIE::CoreOp coreOp) {
+  rewriter.setInsertionPoint(coreOp);
+  auto trueAttr = mlir::BoolAttr::get(rewriter.getContext(), true);
+  auto whileOp = rewriter.create<scf::WhileOp>(
+      rewriter.getUnknownLoc(), TypeRange{}, ValueRange{},
+      [&](OpBuilder &beforeBuilder, Location loc, ValueRange operands) {
+        auto condition = beforeBuilder.create<arith::ConstantOp>(loc, trueAttr);
+        beforeBuilder.create<scf::ConditionOp>(loc, condition, operands);
+      },
+      nullptr);
+  Block *coreBlock = coreOp.getBody();
+  rewriter.eraseOp(coreBlock->getTerminator());
+  Block *afterBlock = whileOp.getAfterBody();
+  rewriter.mergeBlocks(coreBlock, afterBlock);
+  rewriter.setInsertionPoint(afterBlock, afterBlock->end());
+  rewriter.create<scf::YieldOp>(rewriter.getUnknownLoc(), ValueRange{});
+  Block *newBlock = rewriter.createBlock(&coreOp.getRegion());
+  rewriter.setInsertionPoint(newBlock, newBlock->begin());
+  auto endOp = rewriter.create<AMDAIE::EndOp>(rewriter.getUnknownLoc());
+  rewriter.moveOpBefore(whileOp, endOp);
+  return success();
+}
+
+/// Insert a loop with extent equal to the least common multiple of the depths
+/// of the objFifos operated on. This ensures functional correctness after
+/// potential further loop unrolling based on the depths of the objFifos.
+/// Specifically, this anticipates the unrolling happening inside the
+/// `AcquireReleaseToUseLock` pass and is somewhat coupled with it.
```

Comment:
Still not clear to me why this is needed for coreLoopUnroll. Since this pass will be run before `AcquireReleaseToUseLock`, is it possible/more clear to move `coreLoopUnroll` logic to this pass?

Reply:
If I don't insert the loop with extent `lcm(depths)`, then there would be a while loop and unrolling wouldn't be possible, leading to numerical issues as different iterations of the loop potentially need to operate on different buffers (double buffering).

> Since this pass will be run before AcquireReleaseToUseLock, is it possible/more clear to move coreLoopUnroll logic to this pass?

No, because the pass in this PR is only run optionally and `coreLoopUnroll` always needs to be run.